datasquare_test.go

```diff
@@ -471,6 +471,57 @@ func BenchmarkEDSRootsWithErasuredNMT(b *testing.B) {
 	}
 }
 
+func BenchmarkEDSRootsWithBufferedErasuredNMT(b *testing.B) {
+	const mebibyte = 1024 * 1024            // bytes
+	ODSSizeByteUpperBound := 512 * mebibyte // converting 512 MiB to bytes
+	totalNumberOfShares := float64(ODSSizeByteUpperBound) / shareSize
+	// the closest power of 2 of the square root of
+	// the total number of shares
+	nearestPowerOf2ODSSize := math.Pow(2, math.Ceil(math.Log2(math.Sqrt(
+		totalNumberOfShares))))
+	namespaceIDSize := 8
+
+	for squareSize := 32; squareSize <= int(nearestPowerOf2ODSSize); squareSize *= 2 {
+		// number of shares in the original data square's row/column
+		odsSize := squareSize
+		// number of shares in the extended data square's row/column
+		edsSize := 2 * odsSize
+		// generate an EDS with edsSize X edsSize dimensions in terms of shares.
+		// the generated EDS does not conform to celestia-app specs in terms
+		// of namespace version, also no erasure encoding takes place
+		// yet none of these should impact the benchmarking
+		ds := genRandSortedDS(edsSize, shareSize, namespaceIDSize)
+
+		// a tree constructor for erasured nmt
+		parallelOps := 40
```

Comment:
what's the reasoning behind 40? probably we should use the number of CPUs?

Reply:
So in my case it was empirically equal to numCPU * 4. In general we need more than numCPU goroutines because of latency hiding - when a goroutine blocks reading from RAM, the CPU can switch to another goroutine instead of sitting idle. So while ~30 goroutines are blocked waiting for memory, ~10 are actively computing hashes, keeping all CPU cores busy. The 4x multiplier ensures we always have ready goroutines to hide memory latency.